File: game/atc-core/server/lockdown.lua
-- Phase 70: Final Runtime Consolidation, Deterministic Simulation Closure & Production Lockdown

local BASE = 'http://localhost:30120/api/v1/lockdown'

-- Runtime Lockdown
AddEventHandler('atc:lockdown:initiate', function(data, cb)
    PerformHttpRequest(BASE, function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:activate', function(id, cb)
    PerformHttpRequest(BASE .. '/' .. id .. '/activate', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:lift', function(id, cb)
    PerformHttpRequest(BASE .. '/' .. id .. '/lift', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:fail', function(id, cb)
    PerformHttpRequest(BASE .. '/' .. id .. '/fail', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Deterministic Closure
AddEventHandler('atc:lockdown:closure:start', function(data, cb)
    PerformHttpRequest(BASE .. '/closure', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:closure:complete', function(id, cb)
    PerformHttpRequest(BASE .. '/closure/' .. id .. '/complete', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:closure:abort', function(id, cb)
    PerformHttpRequest(BASE .. '/closure/' .. id .. '/abort', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Production Integrity
AddEventHandler('atc:lockdown:integrity:create', function(data, cb)
    PerformHttpRequest(BASE .. '/integrity', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:integrity:pass', function(id, cb)
    PerformHttpRequest(BASE .. '/integrity/' .. id .. '/pass', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:integrity:fail', function(id, cb)
    PerformHttpRequest(BASE .. '/integrity/' .. id .. '/fail', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Runtime Seals
AddEventHandler('atc:lockdown:seal:apply', function(data, cb)
    PerformHttpRequest(BASE .. '/seal', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:seal:verify', function(id, cb)
    PerformHttpRequest(BASE .. '/seal/' .. id .. '/verify', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:seal:break', function(id, cb)
    PerformHttpRequest(BASE .. '/seal/' .. id .. '/break', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Distributed Finalization
AddEventHandler('atc:lockdown:finalization:start', function(data, cb)
    PerformHttpRequest(BASE .. '/finalization', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', json.encode(data), { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:finalization:commit', function(id, cb)
    PerformHttpRequest(BASE .. '/finalization/' .. id .. '/commit', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

AddEventHandler('atc:lockdown:finalization:rollback', function(id, cb)
    PerformHttpRequest(BASE .. '/finalization/' .. id .. '/rollback', function(status, body)
        if cb then cb(status == 200 and json.decode(body) or nil) end
    end, 'POST', '', { ['Content-Type'] = 'application/json' })
end)

-- Cleanup
AddEventHandler('atc:lockdown:cleanup', function(thresholdMs)
    PerformHttpRequest(BASE .. '/cleanup', function() end,
        'POST', json.encode({ thresholdMs = thresholdMs or 300000 }),
        { ['Content-Type'] = 'application/json' })
end)

-- Scheduled cleanup every 5 minutes
CreateThread(function()
    while true do
        Wait(300000)
        TriggerEvent('atc:lockdown:cleanup', 300000)
    end
end)

File: apps/api/src/routes/lockdown.ts
import type { FastifyInstance } from 'fastify'
import type { AppContext } from '../context.js'
import {
  initiateLockdownSchema,
  startClosureSchema,
  createProductionIntegrityCheckSchema,
  applySealSchema,
  startFinalizationSchema,
  cleanupLockdownSchema,
} from '@atc/operations'

export function lockdownRoutes(fastify: FastifyInstance, { ctx }: { ctx: AppContext }) {
  // ── Runtime Lockdown ─────────────────────────────────────────────────────────

  fastify.post('/api/v1/lockdown', async (req, reply) => {
    if (!ctx.runtimeLockdownService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = initiateLockdownSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { lockdownData, ...rest } = parsed.data
    const lockdown = await ctx.runtimeLockdownService.initiateLockdown({
      ...rest,
      ...(lockdownData !== undefined ? { lockdownData } : {}),
    })
    return reply.code(200).send(lockdown)
  })

  fastify.post('/api/v1/lockdown/:id/activate', async (req, reply) => {
    if (!ctx.runtimeLockdownService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const lockdown = await ctx.runtimeLockdownService.activateLockdown(id)
    return reply.code(200).send(lockdown)
  })

  fastify.post('/api/v1/lockdown/:id/lift', async (req, reply) => {
    if (!ctx.runtimeLockdownService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const lockdown = await ctx.runtimeLockdownService.liftLockdown(id)
    return reply.code(200).send(lockdown)
  })

  fastify.post('/api/v1/lockdown/:id/fail', async (req, reply) => {
    if (!ctx.runtimeLockdownService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const lockdown = await ctx.runtimeLockdownService.failLockdown(id)
    return reply.code(200).send(lockdown)
  })

  fastify.get('/api/v1/lockdown/:id', async (req, reply) => {
    if (!ctx.runtimeLockdownService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const lockdown = await ctx.runtimeLockdownService.getLockdown(id)
    if (!lockdown) return reply.code(404).send({ error: 'Lockdown record not found' })
    return reply.code(200).send(lockdown)
  })

  // ── Deterministic Closure ────────────────────────────────────────────────────

  fastify.post('/api/v1/lockdown/closure', async (req, reply) => {
    if (!ctx.deterministicClosureService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = startClosureSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { closureData, ...rest } = parsed.data
    const closure = await ctx.deterministicClosureService.startClosure({
      ...rest,
      ...(closureData !== undefined ? { closureData } : {}),
    })
    return reply.code(200).send(closure)
  })

  fastify.post('/api/v1/lockdown/closure/:id/progress', async (req, reply) => {
    if (!ctx.deterministicClosureService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const closure = await ctx.deterministicClosureService.progressClosure(id)
    return reply.code(200).send(closure)
  })

  fastify.post('/api/v1/lockdown/closure/:id/complete', async (req, reply) => {
    if (!ctx.deterministicClosureService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const closure = await ctx.deterministicClosureService.completeClosure(id)
    return reply.code(200).send(closure)
  })

  fastify.post('/api/v1/lockdown/closure/:id/abort', async (req, reply) => {
    if (!ctx.deterministicClosureService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const closure = await ctx.deterministicClosureService.abortClosure(id)
    return reply.code(200).send(closure)
  })

  fastify.get('/api/v1/lockdown/closure/:id', async (req, reply) => {
    if (!ctx.deterministicClosureService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const closure = await ctx.deterministicClosureService.getClosure(id)
    if (!closure) return reply.code(404).send({ error: 'Closure record not found' })
    return reply.code(200).send(closure)
  })

  // ── Production Integrity ─────────────────────────────────────────────────────

  fastify.post('/api/v1/lockdown/integrity', async (req, reply) => {
    if (!ctx.productionIntegrityService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = createProductionIntegrityCheckSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { integrityData, ...rest } = parsed.data
    const check = await ctx.productionIntegrityService.createIntegrityCheck({
      ...rest,
      ...(integrityData !== undefined ? { integrityData } : {}),
    })
    return reply.code(200).send(check)
  })

  fastify.post('/api/v1/lockdown/integrity/:id/run', async (req, reply) => {
    if (!ctx.productionIntegrityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const check = await ctx.productionIntegrityService.beginRunning(id)
    return reply.code(200).send(check)
  })

  fastify.post('/api/v1/lockdown/integrity/:id/pass', async (req, reply) => {
    if (!ctx.productionIntegrityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const check = await ctx.productionIntegrityService.passIntegrityCheck(id)
    return reply.code(200).send(check)
  })

  fastify.post('/api/v1/lockdown/integrity/:id/fail', async (req, reply) => {
    if (!ctx.productionIntegrityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const check = await ctx.productionIntegrityService.failIntegrityCheck(id)
    return reply.code(200).send(check)
  })

  fastify.get('/api/v1/lockdown/integrity/:id', async (req, reply) => {
    if (!ctx.productionIntegrityService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const check = await ctx.productionIntegrityService.getIntegrityCheck(id)
    if (!check) return reply.code(404).send({ error: 'Production integrity check not found' })
    return reply.code(200).send(check)
  })

  // ── Runtime Seals ────────────────────────────────────────────────────────────

  fastify.post('/api/v1/lockdown/seal', async (req, reply) => {
    if (!ctx.runtimeSealService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = applySealSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { sealData, ...rest } = parsed.data
    const seal = await ctx.runtimeSealService.applySeal({
      ...rest,
      ...(sealData !== undefined ? { sealData } : {}),
    })
    return reply.code(200).send(seal)
  })

  fastify.post('/api/v1/lockdown/seal/:id/verify', async (req, reply) => {
    if (!ctx.runtimeSealService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const seal = await ctx.runtimeSealService.verifySeal(id)
    return reply.code(200).send(seal)
  })

  fastify.post('/api/v1/lockdown/seal/:id/break', async (req, reply) => {
    if (!ctx.runtimeSealService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const seal = await ctx.runtimeSealService.breakSeal(id)
    return reply.code(200).send(seal)
  })

  fastify.get('/api/v1/lockdown/seal/:id', async (req, reply) => {
    if (!ctx.runtimeSealService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const seal = await ctx.runtimeSealService.getSeal(id)
    if (!seal) return reply.code(404).send({ error: 'Runtime seal not found' })
    return reply.code(200).send(seal)
  })

  // ── Distributed Finalization ──────────────────────────────────────────────────

  fastify.post('/api/v1/lockdown/finalization', async (req, reply) => {
    if (!ctx.distributedFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = startFinalizationSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const { finalizationData, ...rest } = parsed.data
    const finalization = await ctx.distributedFinalizationService.startFinalization({
      ...rest,
      ...(finalizationData !== undefined ? { finalizationData } : {}),
    })
    return reply.code(200).send(finalization)
  })

  fastify.post('/api/v1/lockdown/finalization/:id/commit', async (req, reply) => {
    if (!ctx.distributedFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const finalization = await ctx.distributedFinalizationService.commitFinalization(id)
    return reply.code(200).send(finalization)
  })

  fastify.post('/api/v1/lockdown/finalization/:id/rollback', async (req, reply) => {
    if (!ctx.distributedFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const finalization = await ctx.distributedFinalizationService.rollbackFinalization(id)
    return reply.code(200).send(finalization)
  })

  fastify.get('/api/v1/lockdown/finalization/:id', async (req, reply) => {
    if (!ctx.distributedFinalizationService) return reply.code(503).send({ error: 'Service unavailable' })
    const { id } = req.params as { id: string }
    const finalization = await ctx.distributedFinalizationService.getFinalization(id)
    if (!finalization) return reply.code(404).send({ error: 'Finalization record not found' })
    return reply.code(200).send(finalization)
  })

  // ── Cleanup ──────────────────────────────────────────────────────────────────

  fastify.post('/api/v1/lockdown/cleanup', async (req, reply) => {
    if (!ctx.lockdownRecoveryService) return reply.code(503).send({ error: 'Service unavailable' })
    const parsed = cleanupLockdownSchema.safeParse(req.body)
    if (!parsed.success) return reply.code(400).send({ error: parsed.error.flatten() })
    const result = await ctx.lockdownRecoveryService.cleanupStale(parsed.data.thresholdMs)
    return reply.code(200).send(result)
  })
}
